crates/gen-rust/src/lib.rs

```diff
@@ -379,6 +380,87 @@ pub trait RustGenerator {
         return result;
     }
 
+    /// Returns the camel-cased 'name' of the passed type, as used to name union variants.
+    fn name(&self, iface: &Interface, ty: &Type) -> String {
+        match ty {
+            Type::Unit => "Unit".to_owned(),
+            Type::Bool => "Bool".to_owned(),
+            Type::U8 => "U8".to_owned(),
+            Type::U16 => "U16".to_owned(),
+            Type::U32 => "U32".to_owned(),
+            Type::U64 => "U64".to_owned(),
+            Type::S8 => "I8".to_owned(),
+            Type::S16 => "I16".to_owned(),
+            Type::S32 => "I32".to_owned(),
+            Type::S64 => "I64".to_owned(),
+            Type::Float32 => "F32".to_owned(),
+            Type::Float64 => "F64".to_owned(),
+            Type::Char => "Char".to_owned(),
+            Type::String => "String".to_owned(),
+            Type::Handle(id) => iface.resources[*id].name.to_camel_case(),
+            Type::Id(id) => {
+                let ty = &iface.types[*id];
+                match &ty.name {
+                    Some(name) => name.to_camel_case(),
+                    None => match &ty.kind {
+                        TypeDefKind::Option(ty) => format!("Optional{}", self.name(iface, ty)),
+                        TypeDefKind::Expected(_) => "Result".to_owned(),
+                        TypeDefKind::Tuple(_) => "Tuple".to_owned(),
+                        TypeDefKind::List(ty) => format!("{}List", self.name(iface, ty)),
+                        TypeDefKind::Stream(s) => format!("{}Stream", self.name(iface, &s.element)),
```

Comment:
Due to the recursive nature of this method could this method take `&mut String` and the methods push onto it instead of having an owned string returned from each level?